File: tests/manual_test_samples/inefficient.py

import time
import math
import os
import sys

def extremely_inefficient_python():
    print("Running EXTREMELY inefficient Python code...")
    start_time = time.time()
    
    # 1. Very inefficient loop iteration (Pattern: range(len()))
    # Metric: High loop overhead, index lookup cost
    large_list = [i for i in range(50000)]
    result_list = []
    
    print("1. Testing loop inefficiency...")
    for i in range(len(large_list)):
        # Inefficient: retrieving by index instead of iterating directly
        val = large_list[i]
        # Inefficient: append inside loop instead of list comprehension
        result_list.append(val * 2)
        
    # 2. String concatenation hell (Pattern: += in loop)
    # Metric: Huge memory churn creating new string objects
    print("2. Testing string inefficiency...")
    output_string = ""
    words = ["word"] * 5000
    for word in words:
        # Inefficient: creates a new string object every iteration
        output_string += word + " "
        
    # 3. Manual summation (Pattern: loops for math)
    # Metric: Python interpreter overhead vs C-optimized sum()
    print("3. Testing math inefficiency...")
    total = 0
    numbers = [i for i in range(100000)]
    for n in numbers:
        total = total + n
        
    # 4. Nested loops with redundant filtering (Pattern: nested loops)
    # Metric: O(N^2) complexity where O(N) is possible
    print("4. Testing nested loop inefficiency...")
    subset = []
    source_a = [i for i in range(1000)]
    source_b = [i for i in range(1000)]
    
    for a in source_a:
        for b in source_b:
            if a == b:
                subset.append(a)
                
    # 5. Redundant API calls/IO (simulated)
    # Metric: Waiting on IO in serial
    print("5. Testing redundant IO patterns...")
    dummy_files = ["file1", "file2", "file3", "file4", "file5"] * 100
    found_count = 0
    for fname in dummy_files:
        # Inefficient: OS call inside tight loop
        if os.path.exists(fname): 
            found_count += 1
            
    end_time = time.time()
    print(f"Inefficient code finished in {end_time - start_time:.4f} seconds")
    return output_string, total, subset

if __name__ == "__main__":
    extremely_inefficient_python()

File: tests/manual_test_samples/Inefficient.java

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Inefficient {

    public static void main(String[] args) {
        System.out.println("Running inefficient Java code...");
        long startTime = System.currentTimeMillis();

        // 1. Inefficient loop iteration (Pattern: manual for loop)
        List<Integer> list = new ArrayList<>();
        for (int i = 0; i < 50000; i++) {
            list.add(i);
        }

        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < list.size(); i++) {
            // Inefficient: Repeatedly calling .get(i) on list (O(n^2) for LinkedList, adds
            // overhead for ArrayList)
            Integer val = list.get(i);

            // Inefficient: Manual loop logic instead of Stream/parallelStream
            if (val % 2 == 0) {
                result.add(val * 2);
            }
        }

        // 2. String Concatenation in Loop (Pattern: += string)
        String s = "";
        for (int i = 0; i < 2000; i++) {
            // Inefficient: Creates a new StringBuilder and String object each iteration
            s += "number" + i;
        }

        // 3. Excessive Auto-boxing (Pattern: Integer vs int)
        // Metric: Creates unnecessary objects, GC pressure
        Integer sum = 0; // Boxing
        for (int i = 0; i < 100000; i++) {
            sum += i; // Unboxing and re-boxing every iteration
        }

        // 4. Inefficient wrapper class usage
        // Using Boolean instead of boolean
        Boolean flag = Boolean.TRUE;
        if (flag) {
        }

        // 5. Unbuffered I/O (Simulated)
        // Would be System.out.print inside loop without buffer

        long endTime = System.currentTimeMillis();
        System.out.println("Inefficient Java finished in " + (endTime - startTime) + "ms");
    }
}
